src/core/Expression.cc

```diff
@@ -475,77 +475,74 @@ boost::optional<CallableFunction> FunctionCall::evaluate_function_expression(con
   }
 }
 
-struct SimplifiedExpression {
-  const Expression *expression;
-  boost::optional<ContextHandle<Context>> new_context = boost::none;
-  boost::optional<const FunctionCall *> new_active_function_call = boost::none;
-};
-using SimplificationResult = std::variant<SimplifiedExpression, Value>;
+SimplificationResult TernaryOp::simplify_function_body(const std::shared_ptr<const Context>& context) const
+{
+  return SimplifiedExpression{evaluateStep(context)};
+}
+
+SimplificationResult Assert::simplify_function_body(const std::shared_ptr<const Context>& context) const
+{
+  return SimplifiedExpression{evaluateStep(context)};
+}
+
+SimplificationResult Echo::simplify_function_body(const std::shared_ptr<const Context>& context) const
+{
+  return SimplifiedExpression{evaluateStep(context)};
+}
+
+SimplificationResult Let::simplify_function_body(const std::shared_ptr<const Context>& context) const
+{
+  ContextHandle<Context> let_context{Context::create<Context>(context)};
+  let_context->apply_config_variables(*context);
+  return SimplifiedExpression{evaluateStep(let_context), std::move(let_context)};
+}
 
-static SimplificationResult simplify_function_body(const Expression *expression, const std::shared_ptr<const Context>& context)
+SimplificationResult FunctionCall::simplify_function_body(const std::shared_ptr<const Context>& context) const
 {
-  if (!expression) {
+  const auto *call = this;
```

Comment:
Nit: Apart from retaining the old structure, does the call variable make this more readable, or does it makes sense to just use `this`?